### ISSUE TITLE
Port mapping mystery redux

### ISSUE THREAD

**Issue by vfulco** [nginx-proxy/nginx-proxy#309]
Have nginx-proxy running serving a number containers so progress has been made.

A little (noob) flummoxed on situations with many ports involved.  Case in point, a dockerized mail server app (stand alone instructions):

"docker run --privileged -i -t -d -p 25:25 -p 143:143 -p 587:587 \
-v /opt/docker/Data:/etc/pki/tls/mailserver -h hostname.com onlyoffice/mailserver"

Another service requires many, many more ports:

"docker run  -p 25:25  -p 80:80 -p 443:443 -p 110:110 -p 143:143 -p 465:465 -p 587:587 -p 993:993 -p 995:995 -v /your-data-dir/data:/data -t analogic/poste.io"

I've tried multiple versions of "--expose" for every port involved and naming a virtual host and  it still fails to access for mail.foo.asia.  I am unsure if "virtual_port" should be added here but then which port is the correct one?

Separately it could be since the SSL cert for the first app and nginx-proxy need to be labelled differently.  The second app auto-sets SSL certs in the right places.

So bottom line, I am correct to expose all ports even though the instructions suggest only 80 and 443 will be used?  TIA.

**Comment by curtiszimmerman** [comment]
The most important thing to keep in mind here is that Nginx is an HTTP/HTTPS proxy, so it won't work the way you are thinking. This image will not reverse-proxy inbound mail server ports for your containers. 

The way this works out of the box is that your nginx-proxy container is bound to ports 80 and 443 (HTTP and HTTPS). You then bind your HTTP/HTTPS (*web*) services to the nginx-proxy container by using the `-e "VIRTUAL_HOST=foo.com,www.foo.com"` in your container. If your container listens for HTTP/HTTPS requests on a weird port (like Prometheus listening on tcp/9090), and it is using the `EXPOSE <port>` in the Dockerfile, nginx-proxy will automagically detect that port number *if it's the only port* and then bind it to 80/443 based on your config, or the default. So a visitor putting `https://www.foo.com/` into their browser will first get routed to tcp/443 on the nginx-proxy, which then *internally* will route the request to your container on its weird port. So this will "seamlessly" route requests to your container.

Now, let me explain this again using an example. Since this was a particularly hardcore pain point for me over the last year and a half of using this, I'll explain it using a Gitlab image as an example, since it "needs" to have 80, 443, and 22 (ssh) exposed for the UI and SSH access to git. What you do is `-e "VIRTUAL_HOST=gitlab.example.com"` in your `docker run` command and nginx-proxy will magically begin routing web (HTTP/HTTPS) requests for `https://gitlab.example.com` to the Gitlab container. With SSH, you need to explicitly bind a port on the host with `-p 10022:22` (for example).

The crux of this whole post is that *the jwilder/nginx-proxy container does not provide proxy services for any service other than HTTP or HTTPS* because that's the only thing Nginx handles. So you should set up your inbound mail server ports with `--publish` options in your `docker run` command, like `-p 25:25 -p 110:110 -p 465:465 -p 587:587 -p 993:993`. Then *drop* the `-p 80:80 -p 443:443` from your command and instead use `-e "VIRTUAL_HOST=mail.foo.asia"`.

**Comment by wader** [comment]
Maybe possible to create a custom `nginx-proxy` image with some kind of ambassador running?

**Comment by tkw1536** [comment]
Closing because of comment:

> The most important thing to keep in mind here is that Nginx is an HTTP/HTTPS proxy, so it won't work the way you are thinking. This image will not reverse-proxy inbound mail server ports for your containers.
